File: JANL.VB/ExtendedControls/RegexTextBox.vb
Imports System.ComponentModel
Imports System.Drawing
Imports System.Text.RegularExpressions

Public Class RegexTextBox
	Inherits TextBox

	Public Sub New()
		MyBase.New()
		BackColor = Color.White
		ValidColor = Colors.Common.Good
		RequiredColor = Colors.Common.Unknown
		InvalidColor = Colors.Common.Bad
		Validate()
	End Sub

	Protected Overrides Sub OnTextChanged(e As EventArgs)
		IsValid = Validate()
		MyBase.OnTextChanged(e)
	End Sub

	Protected Overrides Sub OnValidated(e As EventArgs)
		MyBase.OnValidated(e)
	End Sub

	Protected Overrides Sub OnValidating(e As CancelEventArgs)
		IsValid = Validate()
		If Not IsValid AndAlso Message.Length > 0 AndAlso Text.Length > 0 Then
			e.Cancel = True
			Msgs.ShowWarning(Message)
		End If

		MyBase.OnValidating(e)
	End Sub

	Private Function Validate() As Boolean
		If Pattern.Length = 0 Then Return True
		If Text.Length = 0 Then
			MyBase.BackColor = If(Required, RequiredColor, BackColor)
			Return Not Required
		End If

		Dim R = Regex.IsMatch(Text, Pattern)
		MyBase.BackColor = If(R, ValidColor, InvalidColor)
		Return R
	End Function

#Region "Properties"
	Private _IsValid As Boolean
	Private _Pattern As String = ""
	Private _Required As Boolean

	<Browsable(False)>
	Public Property IsValid() As Boolean
		Get
			Return _IsValid
		End Get
		Private Set
			_IsValid = Value
		End Set
	End Property

	<Browsable(True), Category("RegexTextBox"), DefaultValue("")>
	Public Property Message As String = ""

	<Browsable(True), Category("RegexTextBox"), DefaultValue("")>
	Public Property Pattern As String
		Get
			Return _Pattern
		End Get
		Set
			_Pattern = Value
			Validate()
		End Set
	End Property

	<Browsable(True), Category("RegexTextBox"), DefaultValue(False)>
	Public Property Required As Boolean
		Get
			Return _Required
		End Get
		Set
			_Required = Value
			Validate()
		End Set
	End Property

#Region "Colors"

	<Browsable(True), Category("RegexTextBox"), DefaultValue(GetType(Color), "0xFFFFFF")>
	Public Shadows Property BackColor As Color

	<Browsable(True), Category("RegexTextBox"), DefaultValue(GetType(Color), "0xFFE4E1")>
	Public Property InvalidColor As Color

	<Browsable(True), Category("RegexTextBox"), DefaultValue(GetType(Color), "0xFFFFE0")>
	Public Property RequiredColor As Color

	<Browsable(True), Category("RegexTextBox"), DefaultValue(GetType(Color), "0x90EE90")>
	Public Property ValidColor As Color

#End Region

#End Region

End Class

File: JANL.WinForms/Controls/RegexTextBox.cs
using JANL.Extensions;
using System;
using System.ComponentModel;
using System.Drawing;
using System.Text.RegularExpressions;
using System.Windows.Forms;

namespace JANL.Controls
{
    /// <summary>
    /// Тестовое поле с проверкой через Regex
    /// </summary>
    public class RegexTextBox : TextBox
    {
        public RegexTextBox()
        {
            BackColor = Color.White;
            ValidColor = Colors.Common.Good;
            RequiredColor = Colors.Common.Unknown;
            InvalidColor = Colors.Common.Bad;
            Validate();
        }

        protected override void OnTextChanged(EventArgs e)
        {
            IsValid = Validate();
            base.OnTextChanged(e);
        }

        protected override void OnValidated(EventArgs e) => base.OnValidated(e);

        protected override void OnValidating(CancelEventArgs e)
        {
            IsValid = Validate();
            if (!IsValid && Message?.Length > 0 && Text.Length > 0)
            {
                e.Cancel = true;
                this.ShowWarning(Message);
            }

            base.OnValidating(e);
        }

        private bool Validate()
        {
            if (Pattern.Length == 0) { return true; }
            if (Text.Length == 0)
            {
                base.BackColor = Required ? RequiredColor : BackColor;
                return !Required;
            }

            var R = Regex.IsMatch(Text, Pattern);
            base.BackColor = R ? ValidColor : InvalidColor;
            return R;
        }

        #region Properties

        private string _Pattern = "";
        private bool _Required;

        /// <summary>
        /// Верно но ли заполнено поле
        /// </summary>
        [Browsable(false)]
        public bool IsValid { get; private set; }

        /// <summary>
        /// Сообщение при неверном заполнении поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(""), Description("Сообщение при неверном заполнении поля")]
        public string Message { get; set; }

        /// <summary>
        /// Шаблон проверки поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(""), Description("Шаблон проверки поля")]
        public string Pattern
        {
            get { return _Pattern; }
            set
            {
                _Pattern = value;
                Validate();
            }
        }

        /// <summary>
        /// Обязательно ли заполнение поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(false), Description("Обязательно ли заполнение поля")]
        public bool Required
        {
            get { return _Required; }
            set
            {
                _Required = value;
                Validate();
            }
        }

        #region Colors

        /// <summary>
        /// Цвет необязательного поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(typeof(Color), "0xFFFFFF"), Description("Цвет необязательного поля")]
        public new Color BackColor { get; set; }

        /// <summary>
        /// Цвет неверно заполненного поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(typeof(Color), "0xFFE4E1"), Description("Цвет неверно заполненного поля")]
        public Color InvalidColor { get; set; }

        /// <summary>
        /// Цвет обязательного поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(typeof(Color), "0xFFFFE0"), Description("Цвет обязательного поля")]
        public Color RequiredColor { get; set; }

        /// <summary>
        /// Цвет верно заполненного поля
        /// </summary>
        [Browsable(true), Category("RegexTextBox"), DefaultValue(typeof(Color), "0x90EE90"), Description("Цвет верно заполненного поля")]
        public Color ValidColor { get; set; }

        #endregion Colors

        #endregion Properties
    }
}